### ISSUE TITLE
Lie detector 

### ISSUE THREAD

**Issue by shacklefordjames60-coder** [mfouad/awesome-egypt-opensource#28]
"""
qsnn_full_pipeline.py
Single-file consolidated QSNN hybrid polygraph toolkit.

Core features:
- QSNNPolygraph (logits output) with device handling
- MC Dropout uncertainty estimation
- Optional PennyLane quantum-forward hook (parameter-shift compatible)
- Temperature scaling calibration (LBFGS)
- ECE reliability diagram plotting (matplotlib)
- AES-GCM encrypted Base64 serialization + decrypt load
- TorchScript export for edge
- Simple replay buffer & safe autonomous refinement
- Demo pipeline under `if __name__ == "__main__":`

WARNING: This system produces probabilistic outputs. Do NOT use it as an automated enforcement tool.
Always require human review for consequential decisions and keep full encrypted audit logs.
"""

import os
import io
import base64
import math
import random
from collections import deque

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F

# Optional dependencies (install if you plan to use the features)
try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    _HAS_CRYPTO = True
except Exception:
    AESGCM = None
    _HAS_CRYPTO = False

# PennyLane is optional; if available, the module will expose a function to build a q-forward
try:
    import pennylane as qml  # type: ignore
    _HAS_PENNYLANE = True
except Exception:
    qml = None
    _HAS_PENNYLANE = False

# Plotting (matplotlib only)
try:
    import matplotlib.pyplot as plt  # noqa: F401
    _HAS_MPL = True
except Exception:
    plt = None
    _HAS_MPL = False

# ---------------------------
# Core QSNN model (logits)
# ---------------------------
class QSNNPolygraph(nn.Module):
    def __init__(self, input_dim=768, hidden_dim=64, num_qubits=4, dropout_p=0.1, device=None):
        super().__init__()
        self.device = device or (torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu"))
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.num_qubits = num_qubits

        # Classical layers
        self.embed = nn.Linear(input_dim, hidden_dim)
        self.dropout = nn.Dropout(dropout_p)

        # Theta param matrix maps hidden -> qubits (learnable)
        self.theta_params = nn.Parameter(torch.randn(hidden_dim, num_qubits) * 0.01)

        # Readout: hidden + quantum scalar -> logit
        self.readout = nn.Linear(hidden_dim + 1, 1)

        # optimizer & loss set outside usually; provide defaults for convenience
        self.to(self.device)

    def quantum_forward_cos(self, h):
        """
        Classical cosine-based simulator: θ = h @ theta_params; q_out = mean(cos(θ)).
        This is a fast differentiable simulator (not hardware-exact).
        """
        thetas = torch.matmul(h, self.theta_params)  # batch x num_qubits
        exp_vals = torch.cos(thetas)  # simulated <Z>
        q_out = exp_vals.mean(dim=1, keepdim=True)  # batch x 1
        return q_out

    def forward(self, x, quantum_forward_fn=None):
        """
        Returns logits (not probabilities). Always return logits for stable BCEWithLogitsLoss.
        quantum_forward_fn: optional function(h) -> q_out (batch x 1). If None, uses internal cos sim.
        """
        x = x.to(self.device)
        h = F.relu(self.embed(x))
        h = self.dropout(h)  # enable MC Dropout if in train mode or force via helper
        q_fn = quantum_forward_fn or self.quantum_forward_cos
        q_out = q_fn(h) if callable(q_fn) else self.quantum_forward_cos(h)
        combined = torch.cat([h, q_out], dim=1)
        logit = self.readout(combined).squeeze(1)  # shape: batch
        return logit

# ---------------------------
# Dummy dataset for demos
# ---------------------------
class DummyDataset(torch.utils.data.Dataset):
    def __init__(self, num_samples=1000, input_dim=768, seed=42):
        rnd = torch.Generator()
        rnd.manual_seed(seed)
        self.inputs = torch.randn(num_samples, input_dim, generator=rnd)
        self.labels = torch.randint(0, 2, (num_samples, 1)).float()
    def __len__(self):
        return len(self.inputs)
    def __getitem__(self, idx):
        return self.inputs[idx], self.labels[idx]

# ---------------------------
# MC Dropout helpers
# ---------------------------
def enable_dropout(model):
    """Enable dropout during inference for MC Dropout (sets dropout modules to train mode)."""
    for m in model.modules():
        if isinstance(m, nn.Dropout):
            m.train()

def mc_dropout_proba(model, x, mc_runs=50, quantum_forward_fn=None):
    """
    Run MC Dropout inference: returns (mean_probs, std_probs)
    model: returns logits
    x: input tensor (batch x input_dim)
    quantum_forward_fn: optional quantum forward to pass through model
    """
    device = model.device
    model.to(device)
    model.eval()
    enable_dropout(model)  # makes dropout stochastic
    x = x.to(device)
    probs = []
    with torch.no_grad():
        for _ in range(mc_runs):
            logits = model(x, quantum_forward_fn=quantum_forward_fn)
            probs.append(torch.sigmoid(logits).cpu().unsqueeze(0))
    probs = torch.cat(probs, dim=0)  # mc x batch
    mean = probs.mean(dim=0)  # batch
    std = probs.std(dim=0)
    return mean, std

# ---------------------------
# PennyLane quantum hook (optional)
# ---------------------------
def make_pennylane_quantum_forward(n_qubits=4, device_name='default.qubit'):
    """
    Build a PennyLane-based quantum forward function.
    Returns: function(h_mapped) -> q_out (batch x 1), where h_mapped is batch x n_qubits (thetas).
    If PennyLane not available, returns None.
    """
    if not _HAS_PENNYLANE:
        print("[qsnn] PennyLane not installed — falling back to cosine sim. pip install pennylane to enable.")
        return None

    dev = qml.device(device_name, wires=n_qubits)

    @qml.qnode(dev, interface='torch', diff_method='parameter-shift')
    def circuit(thetas):
        # thetas: length n_qubits
        for i in range(n_qubits):
            qml.RX(thetas[i], wires=i)
        # small entangling layer
        for i in range(n_qubits - 1):
            qml.CNOT(wires=[i, i+1])
        return [qml.expval(qml.PauliZ(i)) for i in range(n_qubits)]

    def quantum_forward_pennylane(h_mapped):
        batch = h_mapped.shape[0]
        outs = []
        for b in range(batch):
            thetas = h_mapped[b]
            with torch.set_grad_enabled(True):
                res = circuit(thetas)
            outs.append(torch.tensor(res, dtype=torch.float32))
        outs = torch.stack(outs).to(h_mapped.device)
        return outs.mean(dim=1, keepdim=True)

    return quantum_forward_pennylane

# ---------------------------
# Temperature scaling calibration
# ---------------------------
def temperature_scale_logits(val_logits, val_labels, max_iter=200):
    """
    Fit a single temperature scalar T to minimize BCE on validation set.
    val_logits: torch.Tensor (N,) logits
    val_labels: torch.Tensor (N,) 0/1
    Returns T (float)
    """
    device = val_logits.device
    T = torch.ones(1, device=device, requires_grad=True)
    nll = nn.BCEWithLogitsLoss()
    optimizer = optim.LBFGS([T], lr=0.1, max_iter=max_iter)

    def closure():
        optimizer.zero_grad()
        scaled = val_logits / T
        loss = nll(scaled, val_labels)
        loss.backward()
        return loss

    optimizer.step(closure)
    return float(T.detach().cpu().item())

def expected_calibration_error(probs, labels, n_bins=15):
    """
    probs: array-like (N,)
    labels: array-like (N,)
    Returns ECE float and bin summary list
    """
    import numpy as np
    probs = np.asarray(probs)
    labels = np.asarray(labels)
    bins = np.linspace(0.0, 1.0, n_bins + 1)
    ece = 0.0
    bin_summary = []
    for i in range(n_bins):
        lo = bins[i]; hi = bins[i+1]
        # include hi in last bin
        if i == n_bins - 1:
            mask = (probs >= lo) & (probs <= hi)
        else:
            mask = (probs >= lo) & (probs < hi)
        if mask.sum() == 0:
            bin_summary.append((0.0, 0.0, 0))
            continue
        prop_in_bin = mask.mean()
        avg_conf = probs[mask].mean()
        avg_acc = labels[mask].mean()
        ece += abs(avg_conf - avg_acc) * prop_in_bin
        bin_summary.append((float(avg_conf), float(avg_acc), int(mask.sum())))
    return float(ece), bin_summary

def plot_reliability(probs, labels, n_bins=15, title="Reliability Diagram"):
    if not _HAS_MPL:
        print("[qsnn] matplotlib not available; skipping plot.")
        return
    import numpy as np
    probs_np = np.asarray(probs)
    labels_np = np.asarray(labels)
    bins = np.linspace(0.0, 1.0, n_bins + 1)
    bin_centers = 0.5 * (bins[:-1] + bins[1:])
    avg_conf = []
    avg_acc = []
    for i in range(n_bins):
        lo, hi = bins[i], bins[i + 1]
        if i == n_bins - 1:
            mask = (probs_np >= lo) & (probs_np <= hi)
        else:
            mask = (probs_np >= lo) & (probs_np < hi)
        if mask.sum() == 0:
            avg_conf.append(float('nan'))
            avg_acc.append(float('nan'))
        else:
            avg_conf.append(probs_np[mask].mean())
            avg_acc.append(labels_np[mask].mean())
    import matplotlib.pyplot as plt
    plt.figure(figsize=(6, 6))
    plt.plot([0, 1], [0, 1], linestyle='--', label='perfect')
    plt.plot(bin_centers, avg_acc, marker='o', label='accuracy')
    plt.scatter(bin_centers, avg_conf, marker='x', label='confidence')
    plt.xlabel('Confidence')
    plt.ylabel('Accuracy')
    plt.title(title)
    plt.legend()
    plt.grid(True)
    plt.show()

# ---------------------------
# Secure serialization (AES-GCM) and TorchScript export
# ---------------------------
def encrypt_state_dict_b64(model, aes_key=None):
    """
    Returns (b64_payload, key_bytes)
    If AES not available, raises ImportError.
    """
    if not _HAS_CRYPTO:
        raise ImportError("cryptography required for AES-GCM encryption (pip install cryptography)")

    state_buf = io.BytesIO()
    torch.save(model.state_dict(), state_buf)
    raw = state_buf.getvalue()
    if aes_key is None:
        aes_key = AESGCM.generate_key(bit_length=256)
    aesgcm = AESGCM(aes_key)
    nonce = os.urandom(12)
    ct = aesgcm.encrypt(nonce, raw, None)
    payload = nonce + ct
    return base64.b64encode(payload).decode("utf-8"), aes_key

def decrypt_state_dict_b64_to_model(model, b64_str, aes_key):
    if not _HAS_CRYPTO:
        raise ImportError("cryptography required for AES-GCM decryption (pip install cryptography)")
    payload = base64.b64decode(b64_str.encode("utf-8"))
    nonce = payload[:12]
    ct = payload[12:]
    aesgcm = AESGCM(aes_key)
    raw = aesgcm.decrypt(nonce, ct, None)
    buf = io.BytesIO(raw)
    model.load_state_dict(torch.load(buf, map_location="cpu"))
    model.to(model.device)
    return model

def export_torchscript(model, example_input, out_path="qsnn_script.pt"):
    """
    Exports a TorchScript file that returns probability (sigmoid(logits)).
    example_input: tensor shaped (1, input_dim)
    """
    model.eval()
    class Wrapper(nn.Module):
        def __init__(self, m):
            super().__init__()
            self.m = m
        def forward(self, x):
            logits = self.m(x)
            return torch.sigmoid(logits)
    wrapper = Wrapper(model).to(model.device)
    traced = torch.jit.trace(wrapper, example_input.to(model.device))
    traced.save(out_path)
    return out_path

# ---------------------------
# Simple replay buffer
